### PR TITLE
refactor: use middlewares for fetching Chains from database

### PR DESCRIPTION
**Pull request by pitasi**
This is only one attempt to reduce code duplication across api-server codebase, there will be more 😁

This PR adds two middlewares for doing:
1. Get a cns.Chain struct from database
2. Check if the cns.Chain is enabled

I upgraded to Go v1.18 and wrote a small utility func (check `ginutils`) that uses generics.

### CODE REVIEW

File: api/chains/api.go [hunk 1]
```diff
@@ -1,34 +1,120 @@
...
+			c.Abort()
+			return
+		}
+
+		c.Set(ChainContextKey, chain)
+		c.Next()
+	}
+}
+
+// hainMiddleware checks if the chain exists and it's enabled in the database,
```
Comment (gsora) [inline api/chains/api.go]:
```suggestion
// RequireChainEnabled checks if the chain exists and it's enabled in the database,
```

**Review by gsora — APPROVED** [review]
Typo suggestion, but LFTM.

Can't believe I'm approving a PR with generics.

File: api/chains/api.go (hunk 1 above)
Comment (pitasi) [reply api/chains/api.go]: Ops, fixed

**Comment by pitasi** [comment]
> Can't believe I'm approving a PR with generics.

Don't worry I'll find other generics usages 😛

**Review by sahith-narahari — APPROVED** [review]

**Review by lukitsbrian — APPROVED** [review]

**Pull request merged**